installed-tests/js/testGObjectInterface.js

```diff
@@ -0,0 +1,323 @@
+    JSUnit.assertTrue(obj.constructor.implements(MyGObjectInterface));
+    JSUnit.assertTrue(obj.constructor.implements(MyOtherGObjectInterface));
+}
+
+function testGObjectInterfaceCanChainUpToOtherInterface() {
+    let obj = new MyOtherGObject();
+    JSUnit.assertEquals('MyOtherGObjectInterface.optionalG()\nMyGObjectInterface.optionalG()',
+        obj.optionalG());
+}
+
+function testGObjectDefersToLastInterfaceOptionalFunction() {
+    const MyOtherMinimalGObject = new Lang.Class({
+        Name: 'MyOtherMinimalGObject',
+        Extends: GObject.Object,
+        Implements: [ MyGObjectInterface, MyOtherGObjectInterface ],
+        Properties: {
+            'interface-prop': GObject.ParamSpec.override('interface-prop',
+                MyGObjectInterface)
+        },
+
+        _init: function (props={}) {
+            this.parent(props);
+        },
+        requiredG: function () {}
+    });
+    let obj = new MyOtherMinimalGObject();
+    JSUnit.assertEquals('MyOtherGObjectInterface.optionalG()\nMyGObjectInterface.optionalG()',
+        obj.optionalG());
+}
+
+function testGObjectClassMustImplementAllRequiredInterfaces() {
+    JSUnit.assertRaises(() => new Lang.Class({
+        Name: 'MyBadObject',
+        Implements: [ MyOtherGObjectInterface ],
+        required: function () {}
+    }));
+}
+
+function testGObjectClassMustImplementRequiredInterfacesInCorrectOrder() {
+    JSUnit.assertRaises(() => new Lang.Class({
+        Name: 'MyBadObject',
+        Implements: [ MyOtherGObjectInterface, MyGObjectInterface ],
+        required: function () {}
+    }));
+}
+
+function testGObjectInterfaceCanRequireInterfaceFromC() {
+    JSUnit.assertRaises(() => new Lang.Class({
+        Name: 'MyBadObject',
+        Implements: [ MyInitableInterface ]
+    }));
+}
+
+function testGObjectHasInterfaceSignalsAndClassSignals() {
+    let obj = new MyGObject();
+    let interface_signal_emitted = false, class_signal_emitted = false;
+    obj.connect('interface-signal', () => {
+        interface_signal_emitted = true;
+        Mainloop.quit('signal');
+    });
+    obj.connect('class-signal', () => {
+        class_signal_emitted = true;
+        Mainloop.quit('signal');
+    });
+    GLib.idle_add(GLib.PRIORITY_DEFAULT, () => obj.emit('interface-signal'));
+    Mainloop.run('signal');
+    GLib.idle_add(GLib.PRIORITY_DEFAULT, () => obj.emit('class-signal'));
+    Mainloop.run('signal');
+    JSUnit.assertTrue(interface_signal_emitted);
+    JSUnit.assertTrue(class_signal_emitted);
+}
+
+function testGObjectHasInterfacePropertiesAndClassProperties() {
+    let obj = new MyGObject();
+    JSUnit.assertEquals('foobar', obj.interface_prop);
+    JSUnit.assertEquals('meh', obj.class_prop);
+}
+
+// Failing to override an interface property doesn't raise an error but instead
+// logs a critical warning.
+function testGObjectMustOverrideInterfaceProperties() {
+    GLib.test_expect_message('GLib-GObject', GLib.LogLevelFlags.LEVEL_CRITICAL,
```

Comment:
FWIW I don't think it's controversial at all that if a class (not an interface) has some properties, its subclasses also have them. It's the way it works already.
If I read Giovanni's concern correctly, it's that GInterfaces do not come with default getters and setters for properties, so it would be odd to have code to set and get a property with no base class involved; it just turns out to be really convenient to do this in JS.
I think we could take the best of both worlds, and make it so a base class needs to override the properties, but its subclasses should not need to do that, even if they reimplement other parts of the interface.